### PR TITLE
Ruby: Model some ActiveSupport methods

### PR DESCRIPTION
**Pull request by hmac**
Model many ActiveSupport extensions to `String`, `Hash` and `Object`.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveSupport.qll [hunk 1]
```diff
@@ -60,6 +65,90 @@ module ActiveSupport {
...
+          preservesValue = true
+        }
+      }
+
+      private string getExtractComponent(MethodCall mc, int i) {
+        mc.getMethodName() = "extract!" and
+        result = DataFlow::Content::getKnownElementIndex(mc.getArgument(i)).serialize()
+      }
+
+      private class ExtractSummary extends SummarizedCallable {
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/frameworks/ActiveSupport.qll:119]: This summary is sufficiently complicated that I can't quite follow what it's doing. Could you add a comment, perhaps with an example call to `extract!` and then an explanation of the flow step we're adding?

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveSupport.qll [hunk 2]
```diff
@@ -60,6 +65,90 @@ module ActiveSupport {
...
+            "extract!(" +
+              concat(int i, string s | s = getExtractComponent(mc, i) | s, "," order by i) + ")"
+        }
+
+        final override MethodCall getACall() { result = mc }
+
+        override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+          (
+            exists(string s | s = getExtractComponent(mc, _) |
+              input = "Argument[self].Element[" + s + "!]" and
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/frameworks/ActiveSupport.qll:134]: I"m also unsure what the exclamation marks are for.

Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/frameworks/ActiveSupport.qll:134]: See https://github.com/github/codeql/pull/10574 for an explanation.

**Review by alexrford — DISMISSED** [review]
Looks great 👍

**Review by alexrford — APPROVED** [review]
👍

**Pull request merged**